sn_transfers/src/cashnotes/spend_reason.rs

```diff
@@ -89,24 +93,37 @@ impl DiscordName {
         DiscordName {
             hash: Hash::hash(user_name.as_bytes()),
             nonce: DerivationIndex::random(rng),
+            checksum: CHECK_SUM,
         }
     }
 
     fn to_sized_bytes(&self) -> [u8; LIMIT_SIZE] {
         let mut bytes: [u8; LIMIT_SIZE] = [0; LIMIT_SIZE];
         bytes[0..HASH_SIZE].copy_from_slice(self.hash.slice());
-        bytes[HASH_SIZE..LIMIT_SIZE].copy_from_slice(&self.nonce.0);
+        bytes[HASH_SIZE..CONTENT_SIZE].copy_from_slice(&self.nonce.0);
+        bytes[CONTENT_SIZE..LIMIT_SIZE].copy_from_slice(&self.checksum);
         bytes
     }
 
-    fn from_bytes(bytes: &[u8]) -> Self {
+    fn from_bytes(bytes: &[u8]) -> Result<Self> {
         let mut hash_bytes = [0; HASH_SIZE];
         hash_bytes.copy_from_slice(&bytes[0..HASH_SIZE]);
         let hash = Hash::from(hash_bytes.to_owned());
         let mut nonce_bytes = [0; DERIVATION_INDEX_SIZE];
-        nonce_bytes.copy_from_slice(&bytes[HASH_SIZE..LIMIT_SIZE]);
+        nonce_bytes.copy_from_slice(&bytes[HASH_SIZE..CONTENT_SIZE]);
         let nonce = DerivationIndex(nonce_bytes.to_owned());
-        Self { hash, nonce }
+
+        let mut checksum = [0; CHECK_SUM_SIZE];
+        checksum.copy_from_slice(&bytes[CONTENT_SIZE..LIMIT_SIZE]);
+        if checksum != CHECK_SUM {
+            return Err(TransferError::UserNameDecryptFailed);
```

Comment:
Since there's no decryption here, we can straight away report that the key is invalid?
```suggestion
            return Err(TransferError::InvalidDecryptionKey);
```